File: qb-network-graph-be/routes/lineage.py
"""Lineage routes: entities, trail, snapshot, restore."""
from fastapi import APIRouter, Request, Query
from pydantic import BaseModel

router = APIRouter(prefix="/lineage")


class RestoreRequest(BaseModel):
    audit_id: str
    snapshot: dict


@router.get("/entities")
def lineage_entities(request: Request):
    neo4j = request.app.state.neo4j
    return {"data": neo4j.get_lineage_entities()}


@router.get("/trail/{entity_id}")
def audit_trail(request: Request, entity_id: str, limit: int = Query(100)):
    neo4j = request.app.state.neo4j
    return {"data": neo4j.get_audit_trail(entity_id, limit)}


@router.get("/snapshot/{entity_id}")
def entity_snapshot(request: Request, entity_id: str, date: str = Query(...)):
    neo4j = request.app.state.neo4j
    return {"data": neo4j.get_entity_snapshot(entity_id, date)}


@router.post("/restore/{entity_id}")
def restore_entity(request: Request, entity_id: str, body: RestoreRequest):
    neo4j = request.app.state.neo4j
    result = neo4j.restore_entity(entity_id, body.snapshot, body.audit_id)
    return {"data": result}

File: qb-network-graph-ui/src/api/lineage.js
import { apiClient } from './client';

/** Get list of entities that have lineage data. */
export async function getLineageEntities() {
  return apiClient.get('/lineage/entities');
}

/** Get all audit trail entries for an entity, sorted by date ascending. */
export async function getEntityAuditTrail(entityId, limit = 100) {
  return apiClient.get(`/lineage/trail/${entityId}`, { params: { limit } });
}

/**
 * Reconstruct an entity's golden record state at a specific date.
 * Returns the last golden_record_after before the given date.
 */
export async function getEntitySnapshot(entityId, date) {
  return apiClient.get(`/lineage/snapshot/${entityId}`, { params: { date } });
}

/** Restore an entity to a previous golden record state from an audit entry. */
export async function restoreEntity(entityId, auditId, snapshot) {
  return apiClient.post(`/lineage/restore/${entityId}`, { audit_id: auditId, snapshot });
}
